diff --git a/cmake/UserverSetupEnvironment.cmake b/cmake/UserverSetupEnvironment.cmake
index 7f1d52c4e..0e65f749b 100644
--- a/cmake/UserverSetupEnvironment.cmake
+++ b/cmake/UserverSetupEnvironment.cmake
@@ -18,14 +18,6 @@ function(_userver_setup_environment_validate_impl)
 endfunction()
 
 function(_userver_setup_environment_impl)
-    if(USERVER_IMPL_SETUP_ENV_WAS_RUN_FOR_THIS_DIR)
-        return()
-    endif()
-    set(USERVER_IMPL_SETUP_ENV_WAS_RUN_FOR_THIS_DIR
-        ON
-        PARENT_SCOPE
-    )
-
     get_property(USERVER_CMAKE_DIR GLOBAL PROPERTY userver_cmake_dir)
 
     message(STATUS "C compiler: ${CMAKE_C_COMPILER}")
@@ -48,6 +40,14 @@ function(_userver_setup_environment_impl)
         )
     endif()
 
+    if(USERVER_IMPL_SETUP_ENV_WAS_RUN_FOR_THIS_DIR)
+        return()
+    endif()
+    set(USERVER_IMPL_SETUP_ENV_WAS_RUN_FOR_THIS_DIR
+        ON
+        PARENT_SCOPE
+    )
+
     set(CMAKE_EXPORT_COMPILE_COMMANDS
         ON
         PARENT_SCOPE
diff --git a/cmake/modules/FindPostgreSQLInternal.cmake b/cmake/modules/FindPostgreSQLInternal.cmake
index 48a08f023..9ea50d080 100644
--- a/cmake/modules/FindPostgreSQLInternal.cmake
+++ b/cmake/modules/FindPostgreSQLInternal.cmake
@@ -17,6 +17,7 @@ _userver_module_find_include(
     NAMES
     postgres_fe.h
     PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql
     ${USERVER_PG_SERVER_INCLUDE_DIR}
     /usr/include/postgresql/12/server
     /usr/include/postgresql/13/server
@@ -38,6 +39,7 @@ _userver_module_find_include(
     NAMES
     libpq-int.h
     PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/src/interfaces/libpq
     ${USERVER_PG_INCLUDE_DIR}/internal
     ${USERVER_PG_INCLUDE_DIR}/postgresql/internal
     /usr/local/include/postgresql/internal # FreeBSD
@@ -51,6 +53,7 @@ _userver_module_find_include(
     NAMES
     libpq-fe.h
     PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/src/interfaces/libpq
     ${USERVER_PG_INCLUDE_DIR}
     /usr/local/include # FreeBSD
     /usr/local/pgsql/include # postgresql: make install
@@ -60,15 +63,20 @@ _userver_module_find_include(
 )
 
 _userver_module_find_library(
-    NAMES libpq.a # must be a static library as we use internal symbols
-    PATHS ${USERVER_PG_LIBRARY_DIR} /usr/local/lib # FreeBSD
+    NAMES libpostgresql.a # must be a static library as we use internal symbols
+    PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/build/src/interfaces/libpq
+    ${USERVER_PG_LIBRARY_DIR} /usr/local/lib # FreeBSD
     /usr/local/pgsql/lib # postgresql: make install
 )
 
 _userver_module_find_library(
     NAMES
-    libpgcommon.a
+    libpq.so
     PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/build/src/interfaces/libpq
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/build/src/common
     ${USERVER_PG_SERVER_LIBRARY_DIR}
     ${USERVER_PG_LIBRARY_DIR}
     /usr/lib/postgresql/12/lib
@@ -87,6 +95,7 @@ _userver_module_find_library(
     NAMES
     libpgport.a
     PATHS
+    ${CMAKE_SOURCE_DIR}/third_party/postgresql/build/src/port
     ${USERVER_PG_SERVER_LIBRARY_DIR}
     ${USERVER_PG_LIBRARY_DIR}
     /usr/lib/postgresql/12/lib
diff --git a/postgresql/pq-extra/CMakeLists.txt b/postgresql/pq-extra/CMakeLists.txt
index b5bd233d9..1c19de1e6 100644
--- a/postgresql/pq-extra/CMakeLists.txt
+++ b/postgresql/pq-extra/CMakeLists.txt
@@ -8,7 +8,7 @@ add_library(${PROJECT_NAME} STATIC ${PQ_EXTRA_SOURCES})
 
 target_include_directories(${PROJECT_NAME} PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>)
 
-target_link_libraries(${PROJECT_NAME} PUBLIC PostgreSQLInternal userver-internal-compile-options)
+target_link_libraries(${PROJECT_NAME} PUBLIC PostgreSQLInternal userver-internal-compile-options m)
 
 target_compile_options(${PROJECT_NAME} PUBLIC -Wno-ignored-attributes -Wno-pedantic)
 
diff --git a/rabbitmq/CMakeLists.txt b/rabbitmq/CMakeLists.txt
index 5d7829d85..f96bc0a87 100644
--- a/rabbitmq/CMakeLists.txt
+++ b/rabbitmq/CMakeLists.txt
@@ -3,7 +3,7 @@ project(userver-rabbitmq CXX)
 if(USERVER_CONAN)
     find_package(amqpcpp REQUIRED)
 else()
-    include(SetupAmqpCPP)
+    #include(SetupAmqpCPP)
 endif()
 
 userver_module(
diff --git a/third_party/uboost_coro/CMakeLists.txt b/third_party/uboost_coro/CMakeLists.txt
index ba7bc4c58..328b10cdc 100644
--- a/third_party/uboost_coro/CMakeLists.txt
+++ b/third_party/uboost_coro/CMakeLists.txt
@@ -71,7 +71,7 @@ target_compile_options(${PROJECT_NAME} PRIVATE "-w") # no warnings
 target_compile_definitions(${PROJECT_NAME} PRIVATE "BOOST_CONTEXT_SOURCE")
 
 include(Sanitizers)
-_userver_get_sanitize_options(SANITIZERS_LIST SANITIZE_COMPILE_FLAGS SANITIZE_LINK_FLAGS)
+#_userver_get_sanitize_options(SANITIZERS_LIST SANITIZE_COMPILE_FLAGS SANITIZE_LINK_FLAGS)
 
 if("addr" IN_LIST SANITIZERS_LIST)
     target_compile_definitions(${PROJECT_NAME} PUBLIC "BOOST_USE_ASAN")
diff --git a/universal/CMakeLists.txt b/universal/CMakeLists.txt
index c93ca019c..12e82e267 100644
--- a/universal/CMakeLists.txt
+++ b/universal/CMakeLists.txt
@@ -9,7 +9,7 @@ if(TARGET userver-core)
 endif()
 
 include(Sanitizers)
-_userver_make_sanitize_target()
+#_userver_make_sanitize_target()
 
 file(GLOB_RECURSE SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/include/*.hpp ${CMAKE_CURRENT_SOURCE_DIR}/src/*.hpp
      ${CMAKE_CURRENT_SOURCE_DIR}/src/*.cpp
@@ -148,7 +148,7 @@ if(NOT USERVER_NO_WERROR)
     target_compile_options(userver-internal-compile-options INTERFACE "-Werror")
 endif()
 
-_userver_install_targets(COMPONENT universal TARGETS userver-internal-sanitize-options userver-internal-compile-options)
+#_userver_install_targets(COMPONENT universal TARGETS userver-internal-sanitize-options userver-internal-compile-options)
 
 # The library itself
 add_library(${PROJECT_NAME} STATIC ${SOURCES})
@@ -239,7 +239,7 @@ endif()
 
 target_link_libraries(
     ${PROJECT_NAME}
-    PUBLIC Threads::Threads userver-internal-sanitize-options userver-internal-compile-options
+    PUBLIC Threads::Threads userver-internal-compile-options
     PRIVATE Boost::filesystem Boost::program_options OpenSSL::Crypto OpenSSL::SSL
 )
 
